File: integration-tests/src/test/java/com/jaynewstrom/json/sample/retrofit/JsonConverterFactoryTest.java
package com.jaynewstrom.json.sample.retrofit;

import com.fasterxml.jackson.core.JsonFactory;
import com.jaynewstrom.json.retrofit.JsonConverterFactory;
import com.jaynewstrom.json.sample.RealJsonDeserializerFactory;
import com.jaynewstrom.json.sample.RealJsonSerializerFactory;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import java.io.IOException;
import java.util.Collections;
import java.util.List;

import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import retrofit2.Call;
import retrofit2.Converter;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.http.Body;
import retrofit2.http.POST;

import static org.fest.assertions.api.Assertions.assertThat;
import static org.junit.Assert.fail;

public final class JsonConverterFactoryTest {

    interface Service {
        @POST("test") Call<Unknown> callBasicResponseNotSupported();

        @POST("test") Call<Void> callBasicRequestNotSupported(@Body Unknown notSupported);

        @POST("test") Call<Basic> callBasicResponse();

        @POST("test") Call<Void> callBasicRequest(@Body Basic basic);

        @POST("test") Call<List<Basic>> callBasicResponseWithList();

        @POST("test") Call<Void> callBasicRequestWithList(@Body List<Basic> list);

        @POST("test") Call<AutoBasic> callAutoBasicResponse();

        @POST("test") Call<Void> callAutoBasicRequest(@Body AutoBasic basic);
    }

    private static final class Unknown {

    }

    @Rule public final MockWebServer mockWebServer = new MockWebServer();

    private Retrofit retrofit;

    @Before
    public void setUp() {
        retrofit = new Retrofit.Builder()
                .baseUrl(mockWebServer.url("/"))
                .addConverterFactory(converterFactory())
                .build();
    }

    private Service service() {
        return retrofit.create(Service.class);
    }

    private Converter.Factory converterFactory() {
        return JsonConverterFactory.create(new JsonFactory(), new RealJsonSerializerFactory(), new RealJsonDeserializerFactory());
    }

    @Test public void testResponseModelNotSupported() throws InterruptedException, IOException {
        try {
            service().callBasicResponseNotSupported();
            fail();
        } catch (IllegalArgumentException e) {
            assertThat(e).hasMessageContaining("Unable to create converter for class " +
                    "com.jaynewstrom.json.sample.retrofit.JsonConverterFactoryTest$Unknown");
        }
    }

    @Test public void testRequestModelNotSupported() throws InterruptedException, IOException {
        try {
            service().callBasicRequestNotSupported(new Unknown()).execute();
            fail();
        } catch (IllegalArgumentException e) {
            assertThat(e).hasMessageContaining("Unable to create @Body converter for class " +
                    "com.jaynewstrom.json.sample.retrofit.JsonConverterFactoryTest$Unknown");
        }
    }

    @Test public void testResponseConverter() throws InterruptedException, IOException {
        mockWebServer.enqueue(new MockResponse().setBody("{\"foo\":\"bar\"}"));
        Call<Basic> call = service().callBasicResponse();
        Response<Basic> response = call.execute();
        Basic responseBody = response.body();
        assertThat(responseBody.foo).isEqualTo("bar");
    }

    @Test public void testRequestConverter() throws InterruptedException, IOException {
        mockWebServer.enqueue(new MockResponse());
        Basic requestBody = new Basic("bar");
        Call<Void> call = service().callBasicRequest(requestBody);
        call.execute();
        RecordedRequest request = mockWebServer.takeRequest();
        assertThat(request.getHeader("Content-Type")).isEqualTo("application/json; charset=UTF-8");
        assertThat(request.getBody().readUtf8()).isEqualTo("{\"foo\":\"bar\"}");
    }

    @Test public void testResponseConverterWithList() throws InterruptedException, IOException {
        mockWebServer.enqueue(new MockResponse().setBody("[{\"foo\":\"bar\"}]"));
        Call<List<Basic>> call = service().callBasicResponseWithList();
        Response<List<Basic>> response = call.execute();
        List<Basic> responseBody = response.body();
        assertThat(responseBody).hasSize(1);
        assertThat(responseBody.get(0).foo).isEqualTo("bar");
    }

    @Test public void testRequestConverterWithList() throws InterruptedException, IOException {
        mockWebServer.enqueue(new MockResponse());
        Basic requestBody = new Basic("bar");
        Call<Void> call = service().callBasicRequestWithList(Collections.singletonList(requestBody));
        call.execute();
        RecordedRequest request = mockWebServer.takeRequest();
        assertThat(request.getHeader("Content-Type")).isEqualTo("application/json; charset=UTF-8");
        assertThat(request.getBody().readUtf8()).isEqualTo("[{\"foo\":\"bar\"}]");
    }

    @Test public void testResponseConverterUsingAutoValue() throws InterruptedException, IOException {
        mockWebServer.enqueue(new MockResponse().setBody("{\"foo\":\"bar\"}"));
        Call<AutoBasic> call = service().callAutoBasicResponse();
        Response<AutoBasic> response = call.execute();
        AutoBasic responseBody = response.body();
        assertThat(responseBody.getFoo()).isEqualTo("bar");
    }

    @Test public void testRequestConverterUsingAutoValue() throws InterruptedException, IOException {
        mockWebServer.enqueue(new MockResponse());
        AutoBasic requestBody = new AutoValue_AutoBasic("bar");
        Call<Void> call = service().callAutoBasicRequest(requestBody);
        call.execute();
        RecordedRequest request = mockWebServer.takeRequest();
        assertThat(request.getHeader("Content-Type")).isEqualTo("application/json; charset=UTF-8");
        assertThat(request.getBody().readUtf8()).isEqualTo("{\"foo\":\"bar\"}");
    }
}

File: tests/integration-tests/src/test/java/com/jaynewstrom/json/sample/retrofit/JsonConverterFactoryTest.kt
package com.jaynewstrom.json.sample.retrofit

import com.fasterxml.jackson.core.JsonFactory
import com.jaynewstrom.json.retrofit.JsonConverterFactory
import com.jaynewstrom.json.sample.RealJsonDeserializerFactory
import com.jaynewstrom.json.sample.RealJsonSerializerFactory
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.fest.assertions.api.Assertions.assertThat
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import retrofit2.Call
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.http.Body
import retrofit2.http.POST

class JsonConverterFactoryTest {
    @get:Rule val mockWebServer = MockWebServer()

    private lateinit var retrofit: Retrofit

    private interface Service {
        @POST("test") fun callBasicResponseNotSupported(): Call<Unknown>

        @POST("test") fun callBasicRequestNotSupported(@Body notSupported: Unknown): Call<Void>

        @POST("test") fun callBasicResponse(): Call<Basic>

        @POST("test") fun callBasicRequest(@Body basic: Basic): Call<Void>

        @POST("test") fun callBasicResponseWithList(): Call<List<Basic>>

        @POST("test") fun callBasicResponseWithNestedList(): Call<List<List<Basic>>>

        @POST("test") fun callBasicRequestWithList(@Body list: List<Basic>): Call<Void>

        @JvmSuppressWildcards @POST("test") fun callBasicRequestWithNestedList(@Body list: List<List<Basic>>): Call<Void>
    }

    private class Unknown

    @Before
    fun setUp() {
        retrofit = Retrofit.Builder()
            .baseUrl(mockWebServer.url("/"))
            .addConverterFactory(converterFactory())
            .build()
    }

    private fun service(): Service {
        return retrofit.create(Service::class.java)
    }

    private fun converterFactory(): Converter.Factory {
        return JsonConverterFactory.create(JsonFactory(), RealJsonSerializerFactory(), RealJsonDeserializerFactory())
    }

    @Test
    fun testResponseModelNotSupported() {
        try {
            service().callBasicResponseNotSupported()
            fail()
        } catch (e: IllegalArgumentException) {
            assertThat(e).hasMessageContaining("Unable to create converter for class " + "com.jaynewstrom.json.sample.retrofit.JsonConverterFactoryTest\$Unknown")
        }
    }

    @Test
    fun testRequestModelNotSupported() {
        try {
            service().callBasicRequestNotSupported(Unknown()).execute()
            fail()
        } catch (e: IllegalArgumentException) {
            assertThat(e).hasMessageContaining("Unable to create @Body converter for class " + "com.jaynewstrom.json.sample.retrofit.JsonConverterFactoryTest\$Unknown")
        }
    }

    @Test
    fun testResponseConverter() {
        mockWebServer.enqueue(MockResponse().setBody("{\"foo\":\"bar\"}"))
        val call = service().callBasicResponse()
        val response = call.execute()
        val responseBody = response.body()
        assertThat(responseBody!!.foo).isEqualTo("bar")
    }

    @Test
    fun testRequestConverter() {
        mockWebServer.enqueue(MockResponse())
        val requestBody = Basic("bar")
        val call = service().callBasicRequest(requestBody)
        call.execute()
        val request = mockWebServer.takeRequest()
        assertThat(request.getHeader("Content-Type")).isEqualTo("application/json; charset=UTF-8")
        assertThat(request.body.readUtf8()).isEqualTo("{\"foo\":\"bar\"}")
    }

    @Test
    fun testResponseConverterWithList() {
        mockWebServer.enqueue(MockResponse().setBody("[{\"foo\":\"bar\"}]"))
        val call = service().callBasicResponseWithList()
        val response = call.execute()
        val responseBody = response.body()
        assertThat(responseBody).hasSize(1)
        assertThat(responseBody!![0].foo).isEqualTo("bar")
    }

    @Test
    fun testResponseConverterWithNestedList() {
        mockWebServer.enqueue(MockResponse().setBody("[[{\"foo\":\"bar\"}]]"))
        val call = service().callBasicResponseWithNestedList()
        val response = call.execute()
        val responseBody = response.body()
        assertThat(responseBody).hasSize(1)
        assertThat(responseBody!![0]).hasSize(1)
        assertThat(responseBody[0][0].foo).isEqualTo("bar")
    }

    @Test
    fun testRequestConverterWithList() {
        mockWebServer.enqueue(MockResponse())
        val requestBody = Basic("bar")
        val call = service().callBasicRequestWithList(listOf(requestBody))
        call.execute()
        val request = mockWebServer.takeRequest()
        assertThat(request.getHeader("Content-Type")).isEqualTo("application/json; charset=UTF-8")
        assertThat(request.body.readUtf8()).isEqualTo("[{\"foo\":\"bar\"}]")
    }

    @Test
    fun testRequestConverterWithNestedList() {
        mockWebServer.enqueue(MockResponse())
        val requestBody = Basic("bar")
        val call = service().callBasicRequestWithNestedList(listOf(listOf(requestBody)))
        call.execute()
        val request = mockWebServer.takeRequest()
        assertThat(request.getHeader("Content-Type")).isEqualTo("application/json; charset=UTF-8")
        assertThat(request.body.readUtf8()).isEqualTo("[[{\"foo\":\"bar\"}]]")
    }
}
